torch/testing/asserts.py

```diff
@@ -0,0 +1,389 @@
+    """Asserts that both inputs are tensors.
+
+    Args:
+        a (Any): First input.
+        b (Any): Second input.
+
+    Raises:
+        AssertionError: If :attr:`a` or :attr:`b` is not a :class:`~torch.Tensor`.
+    """
+    if not (isinstance(a, torch.Tensor) and isinstance(b, torch.Tensor)):
+        raise AssertionError(f"Both inputs have to be tensors, but got {type(a)} and {type(b)} instead.")
+
+
+def _assert_attributes_equal(
+    a: torch.Tensor,
+    b: torch.Tensor,
+    /,
+    *,
+    check_device: bool = True,
+    check_dtype: bool = True,
+    check_stride: bool = True,
+) -> None:
+    """Asserts that attributes of two tensors match.
+
+    Always checks the :attr:`~torch.Tensor.shape`. Other checks are optional and can be disabled.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+        check_device (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` live in the same
+            :attr:`~torch.Tensor.device` memory.
+        check_dtype (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :attr:`~torch.Tensor.dtype`.
+        check_stride (bool): If ``True`` (default), asserts that both :attr:`a` and :attr:`b` have the same
+            :meth:`~torch.Tensor.stride`.
+
+    Raises:
+        AssertionError: If :attr:`a` and :attr:`b` do not have the same :attr:`~torch.Tensor.shape`.
+        AssertionError: If :attr:`check_device`, but :attr:`a` and :attr:`b` do not live in the same
+            :attr:`~torch.Tensor.device` memory.
+        AssertionError: If :attr:`check_dtype`, but :attr:`a` and :attr:`b` do not have the same
+            :attr:`~torch.Tensor.dtype`.
+        AssertionError: If :attr:`check_stride`, but :attr:`a` and :attr:`b` do not have the same
+            :meth:`~torch.Tensor.stride`.
+    """
+    msg_fmtstr = "The values for attribute '{}' do not match: {} != {}."
+
+    if a.shape != b.shape:
+        raise AssertionError(msg_fmtstr.format("shape", a.shape, b.shape))
+
+    if check_device and a.device != b.device:
+        raise AssertionError(msg_fmtstr.format("device", a.device, b.device))
+
+    if check_dtype and a.dtype != b.dtype:
+        raise AssertionError(msg_fmtstr.format("dtype", a.dtype, b.dtype))
+
+    if check_stride and a.stride() != b.stride():
+        raise AssertionError(msg_fmtstr.format("stride()", a.stride(), b.stride()))
+
+
+def _assert_values_equal(a: torch.Tensor, b: torch.Tensor):
+    """Asserts that the values of two tensors are bitwise equal.
+
+    Args:
+        a (torch.Tensor): First tensor.
+        b (torch.Tensor): Second tensor.
+
+    Raises:
+         AssertionError: If the values of :attr:`a` and :attr:`b` are not bitwise equal.
+
+    .. seealso::
+
+        Internally :func:`torch.eq` is used to check for bitwise equality.
+    """
+    if not torch.all(torch.eq(a, b)):
+        raise AssertionError("ADDME")
```

Comment:
Even for the first PR we should probably put in some reasonable (though minimal) assertion message.